ddtrace/llmobs/_evaluators/ragas/faithfulness.py

```diff
@@ -13,24 +65,236 @@ class RagasFaithfulnessEvaluator:
+        Initialize an evaluator that uses the ragas library to generate a faithfulness score on finished LLM spans.
+
+        Faithfulness measures the factual consistency of an LLM's output against a given context.
+        There are two LLM calls required to generate a faithfulness score - one to generate a set of statements from
+        the answer, and another to measure the faithfulness of those statements against the context using natural
+        language entailment.
+
+        For more information, see https://docs.ragas.io/en/latest/concepts/metrics/faithfulness/
+
+        The `ragas.metrics.faithfulness` instance is used for faithfulness scores. If there is not llm attribute set
+        on this instance, it will be set to the default `llm_factory()` which uses openai.
+
+        :param llmobs_service: An instance of the LLM Observability service used for tracing the evaluation and
+                                submitting evaluation metrics.
+        """
+        self.enabled = True
+        if not RAGAS_DEPENDENCIES_PRESENT:
+            self.enabled = False
+            return
+
         self.llmobs_service = llmobs_service
+        self.ragas_faithfulness_instance = _get_faithfulness_instance()
+
+        if not self.ragas_faithfulness_instance.llm:
+            self.ragas_faithfulness_instance.llm = llm_factory()
+
+        self.llm_output_parser_for_generated_statements = RagasoutputParser(pydantic_object=StatementsAnswers)
+
+        self.llm_output_parser_for_faithfulness_score = RagasoutputParser(pydantic_object=StatementFaithfulnessAnswers)
 
-    def run_and_submit_evaluation(self, span: dict) -> None:
-        if not span:
+        self.split_answer_into_sentences = get_segmenter(
+            language=self.ragas_faithfulness_instance.nli_statements_message.language, clean=False
+        )
+
+    def run_and_submit_evaluation(self, span_event: dict):
+        if not span_event:
             return
-        score_result = self.evaluate(span)
+        score_result = self.evaluate(span_event)
         if score_result:
             self.llmobs_service.submit_evaluation(
-                span_context=span,
+                span_context={"trace_id": span_event.get("trace_id"), "span_id": span_event.get("span_id")},
                 label=RagasFaithfulnessEvaluator.LABEL,
                 metric_type=RagasFaithfulnessEvaluator.METRIC_TYPE,
                 value=score_result,
-                timestamp_ms=math.floor(time.time() * 1000),
             )
 
-    def evaluate(self, span: dict) -> Optional[float]:
-        """placeholder function"""
-        span_id, trace_id = span.get("span_id"), span.get("trace_id")
-        if not span_id or not trace_id:
+    def evaluate(self, span_event: dict) -> Optional[float]:
+        if not self.enabled:
             return None
-        return 1.0
+
+        """Get the latest faithfulness instance from Ragas"""
+        self.ragas_faithfulness_instance = _get_faithfulness_instance()
```

Comment:
How often does RAGAS release a new faithfulness instance and what exactly does this entail? Should we be checking every evaluation?

Reply:
So setting `self.ragas_faithfulness_instance` in `init` to be equal to `ragas.faithfulness` should be enough to ensure the instance has the most up-to-date `llm` and prompt` on every call to `evaluate`

The reason i re-use this function again in `evaluate` is so that `self.ragas_faithfulness_instance.llm` is never `None` ... which only happens in edge cases if the user does something to either accidentally or intentionally set the `faithfulness.llm` instance to None

I can also just remove it and the `evaluate` call can fail silently if there's no LLM for faithfulness, but i thought we should make a best-effort attempt here to still complete the eval successfully

Reply:
I'm fine with that. We should regardless document this behavior both publicly and internally so that it's clear what's happening here